sklearn/metrics/tests/test_pairwise.py

```diff
@@ -606,8 +607,273 @@ def test_cosine_distances():
+
+    assert_array_almost_equal(D_expected, D)
+
+    D = gower_distances(X, categorical_features=[0, 1, 3])
+
+    # These are the normalized values for the initial X above,
+    # but the last column became categorical.
+    X = [['M', False, 0.0, 1],
+         ['F', True, 0.06484477, 2],
+         ['M', True, 1.0, 4],
+         [None, None, np.nan, np.nan]]
+
+    # Simplified calculation of Gower distance for expected values
+    n_rows, n_cols = np.shape(X)
+    D_expected = np.zeros((n_rows, n_rows))
+    for i in range(0, n_rows):
+        for j in range(0, n_rows):
+            D_expected[i][j] = ([1, 0][X[i][0] == X[j][0]] +
+                                [1, 0][X[i][1] == X[j][1]] +
+                                abs(X[i][2] - X[j][2]) +
+                                [1, 0][X[i][3] == X[j][3]]) / n_cols
+
+    D = gower_distances(X, categorical_features=[True, True, False, True],
+                        apply_scaling=False)
+
+    assert_array_almost_equal(D_expected, D)
+
+    # Two observations with same value
+    X = [[1, 4141.22, False, 'ABC'],
+         [1, 4141.22, False, 'ABC']]
+
+    D = gower_distances(X)
+    D_expected = [[0.0, 0.0], [0.0, 0.0]]
+    # An array of zeros is expected as distance, when comparing two
+    # observations with same values.
+    assert_array_almost_equal(D_expected, D)
+
+    # Only categorical values
+    X = [['M', False],
+         ['F', True],
+         ['M', True],
+         ['F', False]]
+
+    X = np.array(X, dtype=np.object)
+
+    # Simplified calculation of Gower distance for expected values
+    n_rows, n_cols = np.shape(X)
+    D_expected = np.zeros((n_rows, n_rows))
+    for i in range(0, n_rows):
+        for j in range(0, n_rows):
+            D_expected[i][j] = ([1, 0][X[i][0] == X[j][0]] +
+                                [1, 0][X[i][1] == X[j][1]]) / n_cols
+
+    D = gower_distances(X)
+
+    assert_array_almost_equal(D_expected, D)
+
+    # Categorical values, with boolean represented as number 1,0
+    X = [['M', 0],
+         ['F', 1],
+         ['M', 1],
+         ['F', 0]]
+
+    D = gower_distances(X, categorical_features=[True, True])
+
+    assert_array_almost_equal(D_expected, D)
+
+    # Categorical values, with boolean represented as 1 and 0,
+    # and missing values
+    X = [['M', 0],
+         ['F', 1],
+         ['M', 1],
+         [None, np.nan]]
+
+    D = gower_distances(X, categorical_features=[True, True])
+
+    n_rows, n_cols = np.shape(X)
+    D_expected = np.zeros((n_rows, n_rows))
+    for i in range(0, n_rows):
+        for j in range(0, n_rows):
+            D_expected[i][j] = ([1, 0][X[i][0] == X[j][0]] +
+                                [1, 0][X[i][1] == X[j][1]]) / n_cols
+
+    # Expected nans need to be put manually, because there is no missing
+    # value treatment in the simplied Gower calculation above.
+    D_expected[:, 3] = [np.nan, np.nan, np.nan, np.nan]
+    D_expected[3] = [np.nan, np.nan, np.nan, np.nan]
```

Comment:
`D_expected[3, :]` may be better for the sake or readability, comparing to the previous line.